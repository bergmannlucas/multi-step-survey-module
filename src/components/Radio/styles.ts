import styled, { css } from 'styled-components'

export const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  margin: 3rem 0;
`

export const OptionWrapper = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  grid-gap: 25px;
`

export const Input = styled.input`
  width: 2rem;
  height: 2rem;
  cursor: pointer;
`

export const Label = styled.span`
  ${({ theme }) => css`
    margin-bottom: 1rem;
    font-size: ${theme.font.sizes.xlarge};
    font-weight: ${theme.font.bold};
    color: ${theme.colors.purple};
  `}
`

export const Option = styled.label`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.xlarge};
    color: ${theme.colors.purple};
    vertical-align: super;
  `}
`
