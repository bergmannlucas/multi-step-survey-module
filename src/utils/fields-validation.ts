import Joi from 'joi'

export const schema = Joi.array().items(
  Joi.object({
    name: Joi.string().empty(''),
    email: Joi.string()
      .email({ tlds: { allow: false } })
      .empty('')
  }),
  Joi.object({
    gender: Joi.string().required(),
    age: Joi.number().required()
  }),
  Joi.object({
    book: Joi.string().required(),
    colors: Joi.object({
      black: Joi.boolean(),
      white: Joi.boolean(),
      red: Joi.boolean(),
      green: Joi.boolean(),
      blue: Joi.boolean(),
      yellow: Joi.boolean(),
      purple: Joi.boolean(),
      orange: Joi.boolean(),
      pink: Joi.boolean()
    })
  })
)
