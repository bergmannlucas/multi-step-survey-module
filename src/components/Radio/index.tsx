import { InputHTMLAttributes } from 'react'

import * as S from './styles'

export type RadioProps = {
  options?: string[] | number[]
  label?: string
  labelFor?: string
  selectedValue?: string
} & InputHTMLAttributes<HTMLInputElement>

function Radio({
  label,
  options,
  labelFor,
  selectedValue,
  ...props
}: RadioProps) {
  return (
    <S.Wrapper>
      {!!label && <S.Label>{label}</S.Label>}
      <S.OptionWrapper>
        {options?.map((option) => (
          <div key={option}>
            <S.Input
              id={labelFor}
              type="radio"
              value={option}
              checked={selectedValue === option}
              onChange={props.onChange}
            />
            <S.Option htmlFor={labelFor}>{option}</S.Option>
          </div>
        ))}
      </S.OptionWrapper>
    </S.Wrapper>
  )
}

export default Radio
