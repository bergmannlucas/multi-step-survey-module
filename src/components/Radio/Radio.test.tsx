import userEvent from '@testing-library/user-event'
import { render, screen, waitFor } from 'utils/test-utils'
import Radio from '.'

describe('<Radio />', () => {
  it('should render with label', () => {
    const { container } = render(
      <Radio
        options={['male']}
        selectedValue="male"
        label="radio label"
        labelFor="radio"
        onChange={jest.fn()}
      />
    )

    expect(screen.getByRole('radio')).toBeInTheDocument()
    expect(screen.getByText(/radio label/i)).toBeInTheDocument()
    expect(screen.getByText(/male/i)).toHaveAttribute('for', 'radio')

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render without label', () => {
    render(<Radio />)

    expect(screen.queryByLabelText('Radio')).not.toBeInTheDocument()
  })

  it('should dispatch onChange when click on radio button', async () => {
    const onChange = jest.fn()
    render(
      <Radio
        label="radio label"
        labelFor="radio"
        onChange={onChange}
        options={['male']}
      />
    )

    expect(onChange).not.toHaveBeenCalled()

    userEvent.click(screen.getByRole('radio'))
    await waitFor(() => {
      expect(onChange).toHaveBeenCalledTimes(1)
    })
  })

  it('should be accessible with tab', () => {
    render(
      <Radio
        label="radio label"
        labelFor="radio"
        options={['male']}
        onChange={jest.fn()}
      />
    )

    const radio = screen.getByRole('radio')

    expect(document.body).toHaveFocus()

    userEvent.tab()

    expect(radio).toHaveFocus()
  })
})
