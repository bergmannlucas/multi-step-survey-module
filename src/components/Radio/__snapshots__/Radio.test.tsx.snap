// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Radio /> should render with label 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: 3rem 0;
}

.c2 {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  grid-gap: 25px;
}

.c3 {
  width: 2rem;
  height: 2rem;
  cursor: pointer;
}

.c1 {
  margin-bottom: 1rem;
  font-size: 2.0rem;
  font-weight: 700;
  color: #9249FF;
}

.c4 {
  font-size: 2.0rem;
  color: #9249FF;
  vertical-align: super;
}

<div
  class="c0"
>
  <span
    class="c1"
  >
    radio label
  </span>
  <div
    class="c2"
  >
    <div>
      <input
        checked=""
        class="c3"
        id="radio"
        type="radio"
        value="male"
      />
      <label
        class="c4"
        for="radio"
      >
        male
      </label>
    </div>
  </div>
</div>
`;
